File: haskell/tests/HelloWorldTests.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Data.Time.Clock (getCurrentTime)
import Data.Time.Format (defaultTimeLocale, formatTime)
import Data.Time.LocalTime (utcToLocalZonedTime)
import Test.Tasty (TestTree, defaultMain, testGroup)
import Test.Tasty.HUnit (assertBool, testCase)

import Hello.World (greeting, printDate, printGreeting)

-- |
-- # @formatErrorMessage@
--
-- The function @formatErrorMessage@ is used to format the error message.
--
-- __Arguments:__
--
-- * @expected: String@ - The expected output.
-- * @actual: String@ - The actual output.
--
-- __Return:__
--
-- * @IO String@ - The formatted error message.
formatErrorMessage :: String -> String -> IO String
formatErrorMessage expected actual = do
    now <- getCurrentTime >>= utcToLocalZonedTime
    return $
        "\nTest run at: "
            ++ show now
            ++ "\nExpected : "
            ++ expected
            ++ "\nActual   : "
            ++ actual
            ++ "\n"

-- |
-- # @printGreetingsWithoutNameTest@
--
-- The function @printGreetingsWithoutNameTest@ is used to test the
-- @printGreeting@ function without args.
printGreetingWithoutNameTest :: TestTree
printGreetingWithoutNameTest = testCase "Print Greeting without Name Test" $ do
    let args = []
    let expectedOutput = "Hello World from Haskell !"
    let actualOutput = printGreeting args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @printGreetingWithNameTest@
--
-- The function @printGreetingWithNameTest@ is used to test the @printGreeting@
-- function with args.
printGreetingWithNameTest :: TestTree
printGreetingWithNameTest = testCase "Print Greeting with Name Test" $ do
    let args = ["Alter", "Sabeh"]
    let expectedOutput = "Hello Alter Sabeh from Haskell !"
    let actualOutput = printGreeting args
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @printDateTest@
--
-- The function @printDateTest@ is used to test the @printDate@ function.
printDateTest :: TestTree
printDateTest = testCase "Print Date Test" $ do
    now <- getCurrentTime >>= utcToLocalZonedTime
    let formattedDate = formatTime defaultTimeLocale "%B %d, %Y at %I:%M %p" now
    let expectedOutput = "Today is " ++ formattedDate
    actualOutput <- printDate
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @printGreetTest@
--
-- The function @printGreetTest@ is used to test the @greeting@ function.
printGreetTest :: TestTree
printGreetTest = testCase "Print Greet Test" $ do
    let expectedOutput = "Hello, Haskell's Here !"
    let actualOutput = greeting
    errorMessage <- formatErrorMessage expectedOutput actualOutput
    assertBool errorMessage (expectedOutput == actualOutput)

-- |
-- # @tests@
--
-- The @tests@ function contains the tests for the @Hello.World@ module.
tests :: TestTree
tests =
    testGroup
        "Hello.World Tests"
        [ printGreetTest
        , printGreetingWithoutNameTest
        , printGreetingWithNameTest
        , printDateTest
        ]

-- |
-- # @main@
--
-- The @main@ function is used to run the tests.
main :: IO ()
main = defaultMain tests

File: rust/tests/hello_world_tests.rs
use chrono::Local;
use colored::Colorize;

use prog::hello_world::{print_date, print_greeting, GREETING};

/// # `format_error_message`
///
/// This function formats the error message.
fn format_error_message(expected: &str, actual: &str) -> colored::ColoredString {
    format!(
        "\n\nTest run at: {}\nExpected : {}\nActual   : {}\n",
        Local::now().to_rfc3339(),
        expected,
        actual
    )
    .red()
}

/// # `print_greeting_without_name_test`
///
/// This function tests the `print_greeting` function without a name.
#[test]
fn print_greeting_without_name_test() {
    let args = vec![];
    let expected_output = "Hello World from Cargo !";
    let actual_output = print_greeting(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `print_greeting_with_name_test`
///
/// This function tests the `print_greeting` function with a name.
#[test]
fn print_greeting_with_name_test() {
    let args = vec!["", "Alter", "Sabeh"];
    let expected_output = "Hello Alter Sabeh from Cargo !";
    let actual_output = print_greeting(&args);
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `print_date_test`
///
/// This function tests the `print_date` function.
#[test]
fn print_date_test() {
    let now = Local::now();
    let formatted_date = now.format("%B %d, %Y at %I:%M %p").to_string();
    let expected_output = format!("Today is {}", formatted_date);
    let actual_output = print_date();
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_eq!(expected_output, actual_output, "{}", error_message);
}

/// # `print_greet_test`
///
/// This function tests the `GREETING` constant.
#[test]
fn print_greet_test() {
    let expected_output = "Hello, Cargo's Here!";
    let actual_output = GREETING;
    let error_message = format_error_message(&expected_output, &actual_output);
    assert_ne!(expected_output, actual_output, "{}", error_message);
}
